File: go/tree/623. 在二叉树中增加一行.go
package tree

func addOneRow(root *TreeNode, v int, d int) *TreeNode {
	if d == 1 {
		node := &TreeNode{Val: v}
		node.Left = root
		return node
	}
	var dfs func(*TreeNode, int)
	dfs = func(root *TreeNode, n int) {
		if root == nil {
			return
		}
		if n+1 == d {
			if root.Left != nil {
				node := &TreeNode{Val: v}
				node.Left = root.Left
				root.Left = node
			} else {
				node := &TreeNode{Val: v}
				root.Left = node
			}
			if root.Right != nil {
				node := &TreeNode{Val: v}
				node.Right = root.Right
				root.Right = node
			} else {
				node := &TreeNode{Val: v}
				root.Right = node
			}
		}
		dfs(root.Left, n+1)
		dfs(root.Right, n+1)
	}
	dfs(root, 1)
	return root
}

File: python/树/623. 在二叉树中增加一行.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def addOneRow(self, root: TreeNode, v: int, d: int) -> TreeNode:
        if d == 1:
            node = TreeNode(v)
            node.left = root
            return node
        def dfs(root,n):
            if not root: return
            if n+1 == d:
                if root.left:
                    node = TreeNode(1)
                    node.left = root.left
                    root.left = node
                else:
                    node = TreeNode(v)
                    root.left = node
                if root.right:
                    node = TreeNode(v)
                    node.right = root.right
                    root.right = node
                else:
                    node = TreeNode(v)
                    root.right = node
            dfs(root.left,n+1)
            dfs(root.right,n+1)
        dfs(root,1)
        return root
